src/ferc1_eia_match/candidate_set_creation.py

```diff
@@ -0,0 +1,344 @@
+        matrix for each dataframe. Optionally set `self.left_blocks_dict` and
+        ``self.right_blocks_dict`` to be the indices of blocks within ``blocking_col``.
+
+        Arguments:
+            blocking_col: The name of the column to get blocks of indices of the
+                dataframes. Default is the empty string if no manual blocking on
+                the value of a column is to be performed. Passing a blocking column
+                will not modify the dataframes themselves, just set the respective
+                dictionary of blocks.
+        """
+        # TODO: there's probably a better way to do this
+        for column_name, vectorizer in self.col_embedding_dict.items():
+            if vectorizer == "tfidf":
+                self.tfidf_vectorize(column_name=column_name)
+            elif vectorizer == "min_max_scale":
+                self.min_max_scale(column_name=column_name)
+        # TODO: allow for other types of aggregation
+        self.left_embedding_matrix = self.equal_weight_aggregate(
+            self.left_embedding_attribute_dict
+        )
+        self.right_embedding_matrix = self.equal_weight_aggregate(
+            self.right_embedding_attribute_dict
+        )
+        if blocking_col != "":
+            self.set_col_blocks(blocking_col=blocking_col)
+
+
+class SimilaritySearcher:
+    """Conduct a search to select a candidate set of likely matched tuples.
+
+    Some methods perform an exact search over all vectors in the index while
+    other searches perform an approximate nearest neighbors search, sacrificing
+    some precision for increased speed.
+
+    FAISS indexes: https://github.com/facebookresearch/faiss/wiki/Faiss-indexes
+    """
+
+    def __init__(self, query_embedding_matrix, menu_embedding_matrix, base_index=None):
+        """Initialize a similarity search object to create a candidate set of tuple pairs.
+
+        If a spare matrix is passed, it will be made dense.
+
+        Arguments:
+            query_embedding_matrix: The embedding matrix representing the query vectors, or
+                vectors for which to search for the k most similar matches from the
+                ``menu_embedding_matrix`` i.e. "for each query vector, find the top k most
+                similar vectors from the index of potential matches".
+            menu_embedding_matrix: The embedding matrix representing the potential match vectors
+                for the query vectors. These vector embeddings which will be searched over to find
+                vectors most similar or closest to the query vectors.
+            base_index (list): The base index for ``menu_embedding_matrix`` that will be
+                searched over for matches. Used to map back to the original vectors if the
+                embeddings were transformed in some way. Default is None and the search
+                will return the position of the vectors in the embedding matrix.
```

Comment:
An example of how I use this `base_index` is in the blocking experiments notebook. Basically it helps to map from the outputted index within the report_year block back to the index of the full dataframe.